File: stacks_and_queues/monotonic/c++/longest_subarray/longest_subarray.cpp
// longest_subarray.cpp
#include "longest_subarray.hpp"
#include <vector>
#include <deque>

int longestSubarray(const std::vector<int> &nums, int limit)
{
    if (nums.size() == 1)
    {
        return 1;
    }

    std::deque<int> monotonicIncreasing;
    std::deque<int> monotonicDecreasing;
    int windowSize = 0, left = 0;

    for (int right = 0; right < nums.size(); right++)
    {
        // While the current element is breaking monotonicities, pop from the deques
        while (!monotonicIncreasing.empty() && monotonicIncreasing.back() > nums[right])
        {
            monotonicIncreasing.pop_back();
        }
        while (!monotonicDecreasing.empty() && monotonicDecreasing.back() < nums[right])
        {
            monotonicDecreasing.pop_back();
        }

        monotonicIncreasing.push_back(nums[right]);
        monotonicDecreasing.push_back(nums[right]);

        // Check if the window has become invalid, i.e., max_of_window - min_of_window > limit
        while (monotonicDecreasing.front() - monotonicIncreasing.front() > limit)
        {
            if (monotonicIncreasing.front() == nums[left])
            {
                monotonicIncreasing.pop_front();
            }
            if (monotonicDecreasing.front() == nums[left])
            {
                monotonicDecreasing.pop_front();
            }
            // Increment the left index to remove it from the window size calculations
            ++left;
        }

        // Check if a larger window size has been found
        windowSize = std::max(windowSize, right - left + 1);
    }

    return windowSize;
}

File: stacks_and_queues/monotonic/python3/longest_subarray.py
from collections import deque
from typing import Deque, MutableSequence


def longest_subarray(nums: MutableSequence[int], limit: int) -> int:
    """
    Given an array of integers `nums` and an integer `limit`, return the size of the longest non-empty subarray such that the absolute difference between
    any two elements of this subarray is less than or equal to `limit`.

    Parameters
    ----------
    nums : MutableSequence[int]
        A mutable sequence of integers
    limit : int
        An integer limit

    Returns
    -------
    int
        The length of the longest continuous valid subarray
    """
    if len(nums) == 1:
        return 1

    monotonic_increasing: Deque[int] = deque()
    monotonic_decreasing: Deque[int] = deque()
    window_size = left = 0
    for right in range(len(nums)):
        # Keep popping until the current element is no longer breaking the monotonic increasing property
        while monotonic_increasing and monotonic_increasing[-1] > nums[right]:
            monotonic_increasing.pop()
        # Keep popping until the current element is no longer breaking the monotonic decreasing property
        while monotonic_decreasing and monotonic_decreasing[-1] < nums[right]:
            monotonic_decreasing.pop()
        # Push the elements
        monotonic_increasing.append(nums[right])
        monotonic_decreasing.append(nums[right])

        # Check if the window has become invalid, i.e., max_of_window - min_of_window > limit
        while monotonic_decreasing[0] - monotonic_increasing[0] > limit:
            # Remove the left most element of the window
            if monotonic_decreasing[0] == nums[left]:
                monotonic_decreasing.popleft()
            if monotonic_increasing[0] == nums[left]:
                monotonic_increasing.popleft()
            # Increment the left pointer now that that element has been removed from the window
            left += 1
        # Check if we have found a longer subarray
        window_size = max(window_size, right - left + 1)

    return window_size


def main() -> int:
    data = [
        ([1, 5, 6, 7, 8, 10, 6, 5, 6], 4),
        ([10, 1, 2, 4, 7, 2], 5),
        ([8, 2, 4, 7], 4),
    ]
    for nums, limit in data:
        window_size = longest_subarray(nums, limit)
        print(
            f"Given nums =  {nums} and limit = {limit}, the longest subarray that satisfies the condition has {window_size} elements"
        )

    return 0


if __name__ == "__main__":
    main()
